src/CacheWithResilience.php

```diff
@@ -0,0 +1,165 @@
+     */
+    public function setItem(CacheItemInterface $item, $value, $ttl): bool
+    {
+        // this condition calculates the DateTime when the element should be consider as expire
+        if ($ttl instanceof DateTimeInterface) {
+            $expiresAt = $ttl;
+        } elseif (\in_array($ttl, [0, -1, CacheInterface::NONE, CacheInterface::INDEFINITE, CacheInterface::SHORT], true)) {
+            // this case handles a list of exception that shouldn't be cached for a different length
+            $ttl = $this->calculateTtl($ttl);
+            $item->expiresAfter($ttl);
+            // getItem() expects the item to be wrapped as [value, ttl]
+            $item->set(['value' => $value, 'ttl' => $ttl]);
+            return $this->cachePool->save($item);
+        } else {
+            $ttl = $this->calculateTtl($ttl);
+            $expiresAt = new DateTime();
+            $expiresAt = $expiresAt->add(new DateInterval('PT' . $ttl . 'S'));
+        }
+
+        // save the ttl within the stored item so we can know if a returned key is expired.
+        $item->set(['value' => $value, 'ttl' => $expiresAt]);
+
+        // instead of setting the item to expire at $expiresAt we use "RESILIENCE_TTL"
+        // this way in case of error we can return the stale value
+        $item->expiresAfter(self::RESILIENCE_TTL);
+
+        return $this->cachePool->save($item);
+    }
+
+    /**
+     * IF CALLABLE RETURNS SOMETHING THAT EVALUATES TO EMPTY THE RESULT WILL NOT BE CACHED UNLESS $nullTtl IS SET
+     * TO A VALUE DIFFERENT FROM CacheInterface::NONE
+     *
+     * @param string $key
+     * @param int|string $ttl
+     *   TTL in seconds, or a constant from CacheInterface
+     * @param callable $function
+     * @param array $arguments
+     * @return mixed
+     */
+    public function getOrSet(string $key, $ttl, callable $function, array $arguments = [], $nullTtl = CacheInterface::NONE)
+    {
+        $cacheItem = $this->getItem($key);
+
+        if ($cacheItem->isHit()) {
+            return $cacheItem->get();
+        }
+
+        try {
+            $result = $function(...$arguments);
+        } catch (\Exception $e) {
+            // if exception is whitelisted, return the stale value
+            if ($this->isWhitelistedException($e)) {
+                $this->logger->error($e->getMessage());
+                // return stale value instead of failing
+                $cacheItem = $this->getItem($key, true);
+                if ($cacheItem->isHit()) {
+                    return $cacheItem->get();
+                }
+            }
+            // if the exception is not whitelisted or the item is not in the cached then throw the expected error.
+            throw $e;
+        }
+
+        if (!empty($result)) {
+            $this->setItem($cacheItem, $result, $ttl);
+        } elseif ($nullTtl !== CacheInterface::NONE) {
+            $this->setItem($cacheItem, $result, $nullTtl);
+        }
+
+        return $result;
+    }
```

Comment:
on the whole, I'm not sure you want the ambiguity of `getOrSet` - better have it do one thing only - either `get` or `set` and have controller logic separately.

Reply:
we have alreadt talk about it but let me know if you want to discuss it any further

Reply:
should be OK, I understand you're keeping the patterns intact. 👍